rust/post-processor/src/processor_status_checker.rs

```diff
@@ -0,0 +1,142 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::metrics::{
+    HASURA_API_LAST_UPDATED_TIME_LATENCY_IN_SECS, HASURA_API_LATEST_VERSION_LATENCY,
+};
+use anyhow::Result;
+use chrono::NaiveDateTime;
+use serde::{Deserialize, Serialize};
+use std::{
+    collections::HashMap,
+    time::{Duration, SystemTime, UNIX_EPOCH},
+};
+use tracing::info;
+
+const PROCESSOR_STATUS_CHECKER_WAIT_TIME_IN_SECS: u64 = 10;
+
+pub struct ProcessorStatusChecker {
+    pub hasura_rest_api_endpoint: String,
+    pub fullnode_rest_api_endpoint: String,
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+struct ProcessorStatusResponse {
+    processor_status: Vec<ProcessorStatus>,
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+struct ProcessorStatus {
+    pub processor: String,
+    pub last_updated: String,
+    pub last_success_version: i64,
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+struct FullnodeResponse {
+    chain_id: u8,
+    epoch: String,
+    ledger_version: String,
+    oldest_ledger_version: String,
+    ledger_timestamp: String,
+    node_role: String,
+    oldest_block_height: String,
+    block_height: String,
+    git_hash: String,
+}
+
+impl ProcessorStatusChecker {
+    pub fn new(hasura_rest_api_endpoint: String, fullnode_rest_api_endpoint: String) -> Self {
+        Self {
+            hasura_rest_api_endpoint,
+            fullnode_rest_api_endpoint,
+        }
+    }
+
+    pub async fn run(&self) -> Result<()> {
+        loop {
+            let processor_latest_version_map = match handle_hasura_response(
+                self.hasura_rest_api_endpoint.clone(),
+            )
+            .await
+            {
+                Ok(result) => result,
+                Err(e) => {
+                    tracing::error!(e = ?e, "Failed to get processor status response from hasura");
+                    panic!();
+                },
+            };
```

Comment:
Cleaner to use `.unwrap_or_else()`. Also maybe add a metric for this failure bc it will just keep on failing and reviving itself in cloud run. Will look something like:
```rust
let processor_latest_version_map = handle_hasura_response().await.unwrap_or_else(|e| {
    error!(e = ?e, "o no");
    FAILURE_METRIC.inc();
    panic!();
});
```

Reply:
Good point. The failure metric is in `main.rs` `TASK_FAILURE_COUNT`.